hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/FrameRunner.java

```diff
@@ -0,0 +1,127 @@
+import com.hedera.node.app.service.contract.impl.hevm.HederaTracer;
+import com.hedera.node.app.service.contract.impl.state.ProxyWorldUpdater;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.processor.ContractCreationProcessor;
+
+/**
+ * An infrastructure service that runs the EVM transaction beginning with the given {@link MessageFrame}
+ * to completion and returns the result.
+ */
+@Singleton
+public class FrameRunner {
+    private final CustomGasCalculator gasCalculator;
+
+    @Inject
+    public FrameRunner(@NonNull final CustomGasCalculator gasCalculator) {
+        this.gasCalculator = gasCalculator;
+    }
+
+    /**
+     * Runs the EVM transaction implied by the given {@link MessageFrame} to completion using the provided
+     * {@link org.hyperledger.besu.evm.processor.AbstractMessageProcessor} implementations, and returns the result.
+     *
+     * @param gasLimit the gas limit for the transaction
+     * @param frame the frame to run
+     * @param tracer the tracer to use
+     * @param messageCall the message call processor to use
+     * @param contractCreation the contract creation processor to use
+     * @return the result of the transaction
+     */
+    public HederaEvmTransactionResult runToCompletion(
+            final long gasLimit,
+            @NonNull final MessageFrame frame,
+            @NonNull final HederaTracer tracer,
+            @NonNull final CustomMessageCallProcessor messageCall,
+            @NonNull final ContractCreationProcessor contractCreation) {
+        requireAllNonNull(frame, tracer, messageCall, contractCreation);
+
+        // We compute the Hedera id up front because the called contract could
+        // selfdestruct, preventing us from looking up its contract id later on
+        final var recipientAddress = frame.getRecipientAddress();
+        final var recipientEvmAddress = asEvmContractId(recipientAddress);
+        final var recipientId = isLongZero(recipientAddress)
+                ? asNumberedContractId(recipientAddress)
+                : ((ProxyWorldUpdater) frame.getWorldUpdater()).getHederaContractId(recipientAddress);
+
+        // Now run the transaction implied by the frame
+        tracer.initProcess(frame);
+        final var stack = frame.getMessageFrameStack();
+        stack.addFirst(frame);
+        while (!stack.isEmpty()) {
+            runToCompletion(stack.peekFirst(), tracer, messageCall, contractCreation);
+        }
+        tracer.finalizeProcess(frame);
+
+        // And package up its result
+        final var gasUsed = effectiveGasUsed(gasLimit, frame);
+        if (frame.getState() == COMPLETED_SUCCESS) {
+            return successFrom(gasUsed, recipientId, recipientEvmAddress, frame);
+        } else {
+            return failureFrom(gasUsed, frame);
+        }
+    }
+
+    private void requireAllNonNull(
```

Comment:
This seems useful and perhaps could be pulled out into a utility function at some point.